## 收集

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## 评估

In [ ]:
patients

In [ ]:
treatments

In [ ]:
adverse_reactions

In [3]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [4]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [5]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [6]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [7]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [8]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [9]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [10]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [11]:
patients.sample(5)

patient_id assigned_sex given_name      surname                address  \
116         117         male    Philipp      Ziegler  3686 Meadowcrest Lane   
184         185       female       Ásta  Grímsdóttir       1619 Melm Street   
34           35       female    Mariana        Souza      577 Chipmunk Lane   
124         125         male       Ming          Hsu   2945 Ferguson Street   
234         235       female    Martina    Tománková                    NaN   

             city state  zip_code        country  \
116         Allen    KY   41601.0  United States   
184  Ormond Beach    FL   32174.0  United States   
34      Orrington    ME    4474.0  United States   
124        Quincy    MA    2169.0  United States   
234           NaN   NaN       NaN            NaN   

                                         contact  birthdate  weight  height  \
116     606-417-4332PhilippZiegler@superrito.com   3/3/1941   143.7      69   
184  AstaGrimsdottir@dayrep.com+1 (386) 989-0019   6/9/1999   128.9      65   
34   207-825-8634MarianaGomesSouza@superrito.com   3/6/1948   152.9      63   
124         MingHsu@armyspy.com+1 (508) 526-3432  11/4/1960   215.6      69   
234                                          NaN   4/7/1936   199.5      65   

      bmi  
116  21.2  
184  21.4  
34   27.1  
124  31.8  
234  33.2

In [38]:
patients.surname.value_counts()[:20]

Doe            6
Jakobsen       3
Taylor         3
Cindrić        2
Aranda         2
Dratchev       2
Collins        2
Souza          2
Liễu           2
Grímsdóttir    2
Bùi            2
Lund           2
Cabrera        2
Tạ             2
Hueber         2
Nilsen         2
Gersten        2
Kowalczyk      2
Schiavone      2
Johnson        2
Name: surname, dtype: int64

In [39]:
patients.address.value_counts()[:20]

123 Main Street               6
2476 Fulton Street            2
2778 North Avenue             2
648 Old Dear Lane             2
4273 Deans Lane               1
2784 Turkey Pen Road          1
77 Juniper Drive              1
522 Lamberts Branch Road      1
3499 Baker Avenue             1
1840 Millbrook Road           1
1589 Roosevelt Wilson Lane    1
1240 Summit Park Avenue       1
1717 Vineyard Drive           1
4535 Prospect Street          1
1846 Joseph Street            1
2235 Catherine Drive          1
2931 Romano Street            1
2386 Linda Street             1
200 Hall Place                1
230 Lady Bug Drive            1
Name: address, dtype: int64

In [14]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [16]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [17]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [18]:
sum(treatments.auralin.isnull())

0

In [19]:
sum(treatments.novodra.isnull())

0

#### 质量
##### `patients` 表格
- 邮编是浮点，而不是字符串
- 邮编有时是四位数
- Tim Neudorf 的身高是 27 英尺，而不是 72 英尺
- 有时使用州的全称，其他用简称
- Dsvid Gustafsson
- 缺少人口统计信息 (地址列至联系方式列) ***(无法清理)***
- 错误的数据类型 (指定性别、州、邮编和出生日期列)
- 多个手机号格式
- John Doe 的默认数据
- Jakobsen、Gersten、Taylor 多条记录
- Zaitseva 体重单位是 kgs，不是 lbs 

##### `treatments` 表格
- 缺少 HbA1c 变化
- Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u' 
- 小写形式的姓和名
- 缺少记录 (280 而不是 350)
- 错误的数据类型 (auralin 和 novodra 列)
- HbA1c 变化不准确 (多个开头的 4 被误写为 9 )
- auralin 和 novodra 列中空值用破折号 (-) 表示

##### `adverse_reactions` 表格
- 小写形式的姓和名

#### 整洁度
-  `patients` 表格中的联系方式一列应该拆分为手机号和邮箱两列
-  `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)
-  不良反应应是 `treatments` 表格的一部分
-  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions` 表格中

## 清理

In [20]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### 缺失数据

<font color='red'>观看 *"首先解决缺失数据问题"* 视频后，完成下列两个 "缺失数据" **定义、代码和测试** 顺序。</font>

#### `treatments`： 缺少记录 (280 而不是 350)

##### 定义
*合并数据。注释：缺少的 `treatments` 记录存储在文件中，命名为 `treatments_cut.csv`，你可以在这个 Jupyter Notebook 指示板中查看 (点击 Notebook 左上角的 **jupyter** 标志)。提示： 解答中使用的函数 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) 。*

##### 代码

In [21]:
# 这里是你的清理代码
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index = True)

##### 测试

In [22]:
# 这里是你的测试代码
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


#### `treatments`：缺少 HbA1c 变化和 HbA1c 变化不准确 (多个开头的 4 被误写为 9)
*注释： "HbA1c 变化不准确 (多个开头的 4 被误写为 9)" 的观察属于准确度问题，而不是完整性问题，由于通过修改缺失的 "缺少 HbA1c 变化" 的清理操作，因此这也被包含在这个标题下面。在一个 **定义、代码和测试** 表头中多个观察数据会在这个记事本中出现。*

##### 定义
*重新计算hba1c_change变量。*

##### 代码

In [23]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [24]:
# 这里是你的清理代码
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end

##### 测试

In [25]:
# 这里是你的测试代码
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

### 整洁度

<font color='red'>观看 *"整洁度清理"* 视频后，完成下列四个 "整洁度" **定义、代码和测试** 顺序。</font>

####  `patients` 表格中的联系方式一列应该包括两个变量：手机号和邮箱

##### 定义
*联系方式分割成邮箱和电话号码。提示 1：通过 pandas 的 [`str.extract` 方法](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html) 使用常规表达。这是个神奇的 [正则表达式教程](https://regexone.com/)。提示 2： [各种手机号正则表达式模式](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number)。提示 3：[邮箱地址正则表达式模式](http://emailregex.com/)，你可能需要用这个进行修改，区分邮箱地址和手机号。*

##### 代码

In [26]:
# 这里是你的清理代码
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [27]:
patients_clean['email'] = pd.Series(patients_clean.contact.str.extract(r'([a-zA-Z\.]+@\w+\.[a-zA-Z]+)'))
patients_clean['phone'] = patients_clean.contact.str.extract(r'([\+1\s]*\(?\d{3}\)?[\s-]\d{3}[\s-]\d{4})')

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


##### 测试

In [29]:
# 这里是你的测试代码
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi                      email              phone  
0  19.6   ZoeWellish@superrito.com       951-719-9170  
1  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2  24.8       JaeMDebord@gustr.com       402-363-6804  
3  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4  26.1        TimNeudorf@cuvox.de       334-515-7487

####  `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)

##### 定义
*添加治疗变量，初始剂量，最终剂量。提示：使用 pandas 的 [融合函数](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) 和 [`str.split()` 方法](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html)。这是非常不错的 [`融合` 教程](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### 代码

In [30]:
treatments_clean = treatments.copy()
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [31]:
# 定义合并auralin和novodra变量函数
def get_outcome(row):
    if row['auralin'] != '-':
        return row['auralin']
    else:
        return row['novodra']

In [32]:
# 将两变量合并
treatments_clean['outcome'] = treatments_clean.apply(get_outcome, axis = 1)

In [33]:
# 使用melt函数将两变量auralin和novodra合并
treatments_clean_outcome = pd.melt(treatments_clean, id_vars = ['given_name'], value_vars = ['auralin', 'novodra'], var_name='treatment', value_name='outcome0')

In [34]:
# 保留有数值的行数
treatments_clean_outcome = treatments_clean_outcome[treatments_clean_outcome.outcome0 != '-']
treatments_clean_outcome.head()

given_name treatment   outcome0
0   veronika   auralin  41u - 48u
3       skye   auralin  33u - 36u
6     sophia   auralin  37u - 42u
7      eddie   auralin  31u - 38u
9       asia   auralin  30u - 36u

In [35]:
# 合并outcome0变量到treatments_clean
treatments_clean = pd.merge(treatments_clean, treatments_clean_outcome, on = 'given_name')
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change    outcome treatment   outcome0  
0           NaN  41u - 48u   auralin  41u - 48u  
1          0.97  40u - 45u   novodra  40u - 45u  
2           NaN  39u - 36u   novodra  39u - 36u  
3          0.35  33u - 36u   auralin  33u - 36u  
4          0.32  33u - 29u   novodra  33u - 29u

In [36]:
# 将outcome0进行拆分
treatments_clean['dose_start'] = treatments_clean.outcome0.str.split('-', n = 2, expand=True)[0]
treatments_clean['dose_end'] = treatments_clean.outcome0.str.split('-', n = 2, expand=True)[1]

##### 测试

In [37]:
# 这里是你的测试代码
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change    outcome treatment   outcome0 dose_start dose_end  
0           NaN  41u - 48u   auralin  41u - 48u       41u       48u  
1          0.97  40u - 45u   novodra  40u - 45u       40u       45u  
2           NaN  39u - 36u   novodra  39u - 36u       39u       36u  
3          0.35  33u - 36u   auralin  33u - 36u       33u       36u  
4          0.32  33u - 29u   novodra  33u - 29u       33u       29u

#### 不良反应应是 `treatments` 表格的一部分

##### 定义
*这里是你的定义。提示：解答中使用的函数 [教程 ](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) 。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

####  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions`小写 表格中，以及小写形式的姓和名 

##### 定义
*这里是你的定义。提示：解答中使用的函数 [教程](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) 和另一个函数的[教程](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) 。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

### 质量

<font color='red'>观看 *"质量清理"* 视频后，完成剩余 "质量" **定义、代码和测试** 顺序。</font>

#### 邮编是浮点，而不是字符串，有时是四位数

##### 定义
*这里是你的定义。提示：查看 "数据清理过程" 页面。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### Tim Neudorf 的身高是 27 英尺，而不是 72 英尺

##### 定义
*这里是你的定义。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### 有时使用州的全称，其他用简称

##### 定义
*这里是你的定义。提示：解答中使用的函数 [教程](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) 。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### Dsvid Gustafsson

##### 定义
*这里是你的定义。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### 错误的数据类型 (指定性别、州、邮编和出生日期列) 和错误的数据类型 (auralin 和 novodra 列)，以及 Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u' 

##### 定义
*这里是你的定义。提示：解答中使用的方法 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html)、函数 [文档页面](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) 以及另一种方法的 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html)。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### 多个手机号格式

##### 定义
*这里是你的定义。提示：有用的 [Stack Overflow 答案](https://stackoverflow.com/a/123681)。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### John Doe 的默认数据

##### 定义
*这里是你的定义。记住，假设替换后的  John Doe 数据不可恢复。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码

#### Jakobsen、Gersten、Taylor 多条记录

##### 定义
*这里是你的定义。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### Test

In [ ]:
# 这里是你的测试代码

#### Zaitseva 体重单位是 kgs，不是 lbs 

##### 定义
*这里是你的定义。*

##### 代码

In [ ]:
# 这里是你的清理代码

##### 测试

In [ ]:
# 这里是你的测试代码